hibernate-core/src/main/java/org/hibernate/sql/results/internal/StandardRowReader.java

```diff
@@ -134,14 +138,22 @@ private void coordinateInitializers(RowProcessingState rowProcessingState) {
 			}
 		}
 
-		for ( int i = 0; i < numberOfInitializers; i++ ) {
+		final Initializer rootInitializer = initializers.get( numberOfInitializers - 1 );
+		rootInitializer.resolveInstance( rowProcessingState );
+
+		if ( !rowProcessingState.isQueryExecution()
```

Comment:
How about we generalize this optimization instead of limiting it just to non-query executions? I'd suggest we add a `default int numberOfNestedInitializers(){ return 0; }` to `Initializer` and use that information to "skip" sub-initializers while traversing the initializers list backwards through something like this:

```
		for ( int i = lastIndex; i >= 0; i-- ) {
			Initializer initializer = initializers.get( i );
			if ( !( initializer instanceof EntityDelayedFetchInitializer ) ) {
				initializer.resolveInstance( rowProcessingState );
				if ( initializer.isInitialized() ) {
					i -= initializer.numberOfNestedInitializers();
				}
			}
		}
```

Reply:
we cannot do it for query 
`select Entity e....`
and then
`select Entity e fetch join e.collection`
and I think we can have the same issue with EntityGraphs

Reply:
for `EntityGraph` my check `!rowProcessingState.isQueryExecution()` is not enough :(

Reply:
Why do you think this is not possible? My intuition is that this can be done, but maybe I'm missing something.